dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp

```diff
@@ -896,22 +886,119 @@ std::string CudaIcoCodeGen::generateStencilInstantiation(
 
   stencilWrapperClass.commit();
 
+  cudaNamespace.commit();
+  dawnNamespace.commit();
+  ssSW << "extern \"C\" {\n";
   bool fromHost = true;
   generateAllAPIRunFunctions(ssSW, stencilInstantiation, codeGenProperties, fromHost);
   generateAllAPIRunFunctions(ssSW, stencilInstantiation, codeGenProperties, !fromHost);
+  ssSW << "}\n";
 
-  cudaNamespace.commit();
-  dawnNamespace.commit();
+  return ssSW.str();
+}
+
+void CudaIcoCodeGen::generateCHeaderSI(
+    std::stringstream& ssSW,
+    const std::shared_ptr<iir::StencilInstantiation>& stencilInstantiation) const {
+  using namespace codegen;
+
+  CodeGenProperties codeGenProperties = computeCodeGenProperties(stencilInstantiation.get());
+
+  ssSW << "extern \"C\" {\n";
+  bool fromHost = true;
+  generateAllAPIRunFunctions(ssSW, stencilInstantiation, codeGenProperties, fromHost,
+                             /*onlyDecl=*/true);
+  generateAllAPIRunFunctions(ssSW, stencilInstantiation, codeGenProperties, !fromHost,
+                             /*onlyDecl=*/true);
+  ssSW << "}\n";
+}
+
+std::string CudaIcoCodeGen::generateCHeader() const {
+  std::stringstream ssSW;
+  ssSW << "#pragma once\n";
+  ssSW << "#include \"driver-includes/defs.hpp\"\n";
+  ssSW << "#include \"driver-includes/cuda_utils.hpp\"\n";
+
+  for(const auto& nameStencilCtxPair : context_) {
+    std::shared_ptr<iir::StencilInstantiation> stencilInstantiation = nameStencilCtxPair.second;
+    generateCHeaderSI(ssSW, stencilInstantiation);
+  }
 
   return ssSW.str();
 }
 
+inline void
+generateF90InterfaceSI(FortranInterfaceModuleGen& fimGen,
+                       const std::shared_ptr<iir::StencilInstantiation>& stencilInstantiation) {
+  const auto& stencils = stencilInstantiation->getStencils();
+
+  // The following assert is needed because we have only one (user-defined) name for a stencil
+  // instantiation (stencilInstantiation->getName()). We could compute a per-stencil name (
+  // codeGenProperties.getStencilName(StencilContext::SC_Stencil, stencil.getStencilID()) ) however
+  // the interface would not be very useful if the name is generated.
+  DAWN_ASSERT_MSG(stencils.size() == 1,
+                  "Unable to generate interface. More than one stencil in stencil instantiation.");
+
+  const auto& stencil = *stencils[0];
+
+  std::vector<FortranInterfaceAPI> apis = {
+      FortranInterfaceAPI("run_" + stencilInstantiation->getName(),
+                          FortranInterfaceAPI::InterfaceType::DOUBLE),
+      FortranInterfaceAPI("run_" + stencilInstantiation->getName() + "_from_fort_host",
+                          FortranInterfaceAPI::InterfaceType::DOUBLE)};
+  for(auto&& api : apis) {
+    api.addArg("mesh", FortranInterfaceAPI::InterfaceType::OBJ);
+    api.addArg("k_size", FortranInterfaceAPI::InterfaceType::INTEGER);
+    for(auto field : support::orderMap(stencil.getFields())) {
+      int n = 3;
```

Comment:
If you merge master you'll find that the field dimensions now offer a function to give you the spatial dimensions of a field. Maybe you can use this to simplify the code at hand